### PR TITLE
Added display of `null` in data table view

### PR DESCRIPTION
**Pull request by ostephens**
As per #1541

### CODE REVIEW

**Comment by ettorerizza** [comment]
The result looks very good! Perhaps we should now use a slightly bolder font for non-zero values, but it's a detail.

screenshot-127 0 0 1-3333-2018 03 17-18-21-34

**Comment by ostephens** [comment]
@jackyq2015 styling invisible characters is separate issue #1286. null is not an invisible character in that sense - it is a cell value not a character

The character issue is more complex and better to keep separate IMO

**Comment by ostephens** [comment]
V happy to look #1546 but prefer to keep tgat separate to this issue & PR

**Comment by wetneb** [comment]
argh - no option to deactivate this? are we 100% sure about that?

**Comment by ettorerizza** [comment]
Rather than an option to disable the display of `null`, why not an option to use empty strings instead of `null ` (basically, a shortcut to Transform all -> `if(isNull(value), "", value)` ? I feel that it is good practice that the users always know when they're working with columns containing null -  that cannot be concatenated directly to one another, unlike columns containing empty strings.

**Comment by thadguidry** [comment]
@ostephens whoa, wait a sec... this was supposed to be an option, as @wetneb said.  We don't want to make this default...its visually too busy for regular users.  Put more work into this and make it an option in our preferences.vt

**Comment by ostephens** [comment]
OK - Apologies for jumping the gun @thadguidry @wetneb. Is the best approach to revert this PR merge and then do some further work for a new PR?

**Comment by wetneb** [comment]
@ostephens I don't think you need to revert this, I think a toggle could just be built on top of that with little effort. It can be quite minimal: just a button that, when clicked, toggles a class on a div that contains the grid, and a CSS rule that puts the nulls in `display: none` when the parent div has that class.

I would be in favour of making this feature opt-in for the coming release (so, disabled by default). Based on user feedback we could consider enabling it by default in the following release.

Basically I'm all in for a better treatment of nulls and empty strings. This is actually something that would be highly beneficial to the Wikidata extension. But I fear that some users might be used to the way things currently are - we don't want to "brutalize" them by forcing on them this distinction that they not might need in their own workflows.

@ettorerizza I'd be careful with such a transformation - it would break the records mode pretty badly, no?

**Comment by ostephens** [comment]
OK - no problem. Will have a look at it hopefully this weekend

**Comment by ettorerizza** [comment]
> I'd be careful with such a transformation - it would break the records mode pretty badly, no?

@wetneb That's what I thought too, but after a few tests, it seems that records can be created with any blanks, nulls or empty strings. 

screencast

In fact, it's not clear to me how exactly records are created and how they can be broken. I mean, I know the big picture of course, but some details are confusing. And there is no documentation on the Wiki. The only mention of the row-records difference I found is on a [Martin's blog post](http://kb.refinepro.com/2012/03/difference-between-record-and-row.html) (@magdmartin : do you think it could be a good idea to copy-paste it on the Wiki?). We will probably have to dive into the source code to flesh out the doc.

**Comment by jackyq2015** [comment]
I took a look of record Modle when I fixed one bug. Here is a simple one:
when column can group together, we have the keyedGroup. Based on the keyedGroup, rowDependency is generated
by searching  the last non-blank cell at following rows. The record model is build upon the rowDependency.
In another word, if there is no rowDependency, record model is meaningless. That is exactly the case when 
importing a FLAT csv file. record model is more suitable for a tree based data structure.

**Comment by wetneb** [comment]
@ettorerizza I agree that the documentation of the records mode is very poor (which is sad because it's quite a fundamental feature) - thankfully there's @magdmartin post but that would clearly deserve to be treated in the wiki. I agree with your analysis of the code - I thought I had read in the past that it relied specifically on nulls but that's apparently wrong. I think that if empty and nulls are conflated in such a crucial part of the code, it's yet another reason not to precipitate displaying the distinction to users…

**Comment by ettorerizza** [comment]
@jackyq2015 The part I don't understand and that makes me feel dumb is : how can the `ExpressionUtils.isNonBlankData() ` method be used to identify both nulls and empty strings (eg in the GREL function [`isBlank()`](https://github.com/OpenRefine/OpenRefine/blob/master/main/src/com/google/refine/grel/controls/IsBlank.java)) while its definition seems to exclude empty strings? (isNonBlankData = non null and not an error and (not a string but possibly a string greater than 0)
